src/common/protocol.ts

```diff
@@ -150,6 +150,54 @@ export class Protocol {
 		}
 	}
 
+	toString(): string {
+		// based on Uri scheme for SSH https://tools.ietf.org/id/draft-salowey-secsh-uri-00.html#anchor1 and heuristics of how GitHub handles ssh url
+		// sshUri        = `ssh:`
+		//    - omitted
+		// hier-part     =  "//" authority path-abempty
+		//    - // is omitted
+		// authority     = [ [ ssh-info ] "@" host ] [ ":" port]
+		//   - ssh-info: git
+		//   - host: ${this.host}
+		//   - port: omitted
+		// path-abempty  = <as specified in [RFC3986]>
+		//   - we use relative path here `${this.owner}/${this.repositoryName}`
+		if (this.type === ProtocolType.SSH) {
+			return `git@${this.host}:${this.owner}/${this.repositoryName}`;
+		}
+
+		if (this.type === ProtocolType.GIT) {
+			return `git://git@${this.host}:${this.owner}/${this.repositoryName}`;
+		}
+
+		let normalizedUri = this.normalizeUri();
+		if (normalizedUri) {
+			return normalizedUri.toString();
+		}
+
+		return null;
+	}
+
+	with(change: { type?: ProtocolType; host?: string; owner?: string; repositoryName?: string; }): Protocol {
```

Comment:
I would call this method `update`. `with` suggests to me an operation that returns a new Protocol without mutating this one.

Reply:
that's right. I didn't make up my mind whether we should make it mutable or not but let's change it to `update` for now. Thanks!